File: clojure/euler/04-palindrome.clj
#!/usr/bin/env clj
;
;  04-palindrome.py
;  code
;
;  Created by Lars Yencken on 2012-01-17.
;  Copyright 2012 Lars Yencken. All rights reserved.
;

(ns euler
  (:use clojure.contrib.combinatorics)
  (:use clojure.test))

(defn is-palindrome? [x]
  (let [s (str x)]
    (= s (apply str (reverse s)))))

(deftest is-palindrome-1
         (is (is-palindrome? 101)))

(deftest is-palindrome-2
         (is true (is-palindrome? 20)))

(deftest is-palindrome-3
         (is true (is-palindrome? 403304)))

(run-tests)

(println)
(println
  "Answer:"
  (reduce max
    (filter #(is-palindrome? %)
            (map #(* (first %) (second %))
                 (combinations (range 100 1000) 2)))))

File: python/euler/04-palindrome.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#  04-palindrome.py
#  code
#
#  Created by Lars Yencken on 2012-01-24.
#  Copyright 2012 Lars Yencken. All rights reserved.
#

"""
Find the largest palindrome made from the product of two 3-digit numbers.
"""

def is_palindrome(x):
    x = str(x)
    return x == x[::-1]

def iter_palindromes():
    for i in xrange(100, 1000):
        for j in xrange(100, 1000):
            x = i * j
            if is_palindrome(x):
                yield x

print max(iter_palindromes())
